platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,895 @@
+    - handled in consensus order
+    - insert into the TSS data map if it is legal to do so
+        - don’t insert multiple messages for the same node
+        - don’t insert any messages if voting window has closed
+    - if inserted into the TSS data map, ensure that the message is forwarded to the TSS Message Validator
+    - if it is a self TSS message, forward the message to the TSS Message creator.
+2. Handle `TssVote` system transactions
+    - handled in consensus order
+    - insert into TSS data map if it is legal to do so
+        - don’t insert multiple votes for the same node
+        - don’t insert once voting is closed
+    - after insertion, re-tally votes and decide if voting is now closed
+    - if voting is closed as the result of a new vote,
+        - if the “no” votes win, then the candidate roster will never be adopted
+        - if the “yes” votes win, the candidate roster is now officially confirmed, and will be adopted at the next
+          upgrade boundary
+3. When a new candidate roster is set.
+    - inform the TSS message creator that there is a new candidate roster
+    - inform the TSS message validator that there is a new candidate roster
+4. Update Metrics
+    - indicate whether there is a candidate roster
+    - the current number of TSS messages collected for the candidate roster
+    - the current number of votes collected for the candidate roster
+        - keep different metrics both for “no” votes and “yes” votes
+5. On First Round After Restart (After PCES Replay and Reconnect)
+    - if there is no candidate roster, take no action
+    - if voting is closed, take no action
+    - if this node’s TSSMessages are not recorded in the TSSData map, send a message to the TSS message creator (so
+      it can resubmit the node’s TSS message)
+    - if this node’s vote is not recorded in the TSS data map, inform the TSS message validator that we still need to do
+      validation.
+
+#### TSS Message Creator
+
+The TSS message creator is responsible for generating a node’s TSSMessages. This runs as its own asynchronous thread
+since this may be a computationally expensive operation.
+
+Internal Elements:
+
+1. TSS Message Manager
+2. active roster field
+3. candidate roster field
+
+Inputs:
+
+1. Active Roster (Constructor)
+2. Candidate Roster (Wire)
+3. Self TSS Message (Wire)
+4. Voting Closed Notification (Wire)
+
+Input Invariants:
+
+1. Self TSS Messages must be received after the candidate roster is set.
+2. The candidate roster is not set if the voting is already closed.
+
+`On Active Roster`:
+
+1. Set this roster as the active roster.
+    1. If the candidate roster is set and the candidate roster is not the active roster, then (re)start the TSS Message
+       Manager with the active roster and candidate roster.
+    2. If the candidate roster is set and the candidate roster is the active roster, then clear the candidate roster,
+       clear the TTS Message List, and stop the TSS Message Manager
```

Comment:
I couldn't find a way to edit your diagram and it was out of date.   
* We were no longer storing anything on disk, bit vector voting replaced the disk.  
* We are not doing state signing through TSS.  
* Your diagram also does not have active or candidate rosters on wires.  

Initially I had a `TssSigningManager` with a generic `ledgerSign(byte[] data)` api for the app to use to generate ledger signatures.  This component replaced the `TssKeyManager` in your diagram.    `TssMessages` were input  to the `TssSigningManager` and nothing else had to get access to the shares from the key material.     The `TssSigningManager` was moved to the `TSS-Block-Signing proposal`. 

That leaves a blank spot where the `TssKeyManager` lives in the earlier diagram. 

Having the active roster on the input wire instead of in the constructor means knowing the wrapping data structure for performing a live roster change.  That would be pulling too much of DAB phase 3 into this effort.   Agreed, doing TSS now means the technical debt for DAB phase 3 increases.   This isn't public API or protobuf, so the platform team has the freedom to evolve the design internally.   DAB phase 3 has to be replanned and go through these platform proposals, anyway.

Reply:
I think I see the problem you're pointing at.  I will sleep on it and probably re-add the `TssKeyManager` back in.